src/runner/mod.rs

```diff
@@ -0,0 +1,190 @@
+            }
+        }
+    }
+
+    fn prepare_before_run(&self) {
+        panic::set_hook(Box::new(|_panic_info| {
+            // XXX panics already catched at the test call site, don't output the trace in stdout
+        }));
+    }
+
+    fn clean_after_run(&self) {
+        // XXX reset panic hook back to default hook:
+        let _ = panic::take_hook();
+    }
+}
+
+impl<T> Visitor<Suite<T>> for Runner
+where
+    T: Clone + Send + Sync + ::std::fmt::Debug,
+{
+    type Environment = T;
+    type Output = ContextReport;
+
+    fn visit(&self, suite: &Suite<T>, environment: &mut Self::Environment) -> Self::Output {
+        self.broadcast(Event::EnterSuite(suite.info.clone()));
+        let report = self.visit(&suite.context, environment);
+        self.broadcast(Event::ExitSuite(report.clone()));
+        report
+    }
+}
+
+impl<T> Visitor<ContextMember<T>> for Runner
+where
+    T: Clone + Send + Sync + ::std::fmt::Debug,
+{
+    type Environment = T;
+    type Output = ContextReport;
+
+    fn visit(&self, member: &ContextMember<T>, environment: &mut Self::Environment) -> Self::Output {
+        match member {
+            &ContextMember::Example(ref example) => {
+                self.visit(example, environment).into()
+            }
+            &ContextMember::Context(ref context) => {
+                self.visit(context, &mut environment.clone())
+            }
+        }
+    }
+}
+
+impl<T> Visitor<Context<T>> for Runner
+where
+    T: Clone + Send + Sync + ::std::fmt::Debug,
+{
+    type Environment = T;
+    type Output = ContextReport;
+
+    fn visit(&self, context: &Context<T>, environment: &mut Self::Environment) -> Self::Output {
+        if let Some(ref info) = context.info {
+            self.broadcast(Event::EnterContext(info.clone()));
+        }
+        for before_function in context.before_all.iter() {
+            before_function(environment);
+        }
+        let process_member = |runner: &Runner, member: &ContextMember<T>, mut environment: T| {
+                for before_each_function in context.before_each.iter() {
+                    before_each_function(&mut environment);
+                }
+            let report = runner.visit(member, &mut environment);
+                for after_each_function in context.after_each.iter() {
+                    after_each_function(&mut environment);
+                }
+                report
+        };
+        let report: ContextReport = if self.configuration.parallel {
+            context
+                .members
+                .par_iter()
+                .map(|member| {
+                    process_member(self, member, environment.clone())
+            })
+                .sum()
+        } else {
+            context
+                .members
+                .iter()
+                .map(|member| {
+                    process_member(self, member, environment.clone())
+                })
+                .sum()
+        };
```

Comment:
It needs to extract the if 

Can the iterator creation be extracted in a function like that:

```rust
let members = self.get_members(); // an iterator that may be parallel depending on the conf
let report = self.process_all_members(members, environment); // the map(process_member) + sum
```